### ISSUE TITLE
Low wall in Fortress of Ice

### ISSUE THREAD

**Issue by Allofich** [afritz1/OpenTESArena#90]
I noticed the left wall right at the entrance of the Fortress of Ice doesn't reach all the way to the ceiling in OpenTESArena. I went and checked in the original and the wall seems to go up all the way to the ceiling.

![opentesarena](https://user-images.githubusercontent.com/19624336/34355563-ae49bff0-ea79-11e7-9b71-1de3dcf0bf78.jpg)
![originalarena](https://user-images.githubusercontent.com/19624336/34355566-b1785510-ea79-11e7-8adf-bad164f70239.jpg)

**Comment by Allofich** [comment]
Comparing the wall textures, it looks like they match at the bottom. But the texture cuts off at the top in OpenTESArena where the wall ends, whereas it continues up for a few more texture pixels up to the ceiling in the original.

**Comment by afritz1** [comment]
Yep, it's a known issue. Those types of voxels are called raised platforms, and they have a Y size and offset from the floor ([definition](https://github.com/afritz1/OpenTESArena/blob/5fce7ee0ae509858f72f3ca0f04b0b408b70bdc3/OpenTESArena/src/World/VoxelData.h#L47)). I think @Carmina16 said that there's a table of values somewhere in the executable where the correct Y values are obtained from. For now, it's just using values from the MAP1 voxel data [here](https://github.com/afritz1/OpenTESArena/blob/1245f46086e9644645a11c17aa69ea1f70376b5c/OpenTESArena/src/World/LevelData.cpp#L548) (as you can see, there are a couple to do's there. I think the voxel loading code could do with another pass).

**Comment by Allofich** [comment]
> I think @Carmina16 said that there's a table of values somewhere in the executable where the correct Y values are obtained from.

@Carmina16, I don't mean to pester you, but is what afritz1 says here true? Maybe you can provide some help for this issue?

**Comment by Carmina16** [comment]
Yes, there is a number of arrays which are referred by those boxes. I gave their coordinates somewhere in the `CEILING` thread. I believe two of them are for the box elevation and thickness, but I didn't check that _in vivo_.

**Comment by afritz1** [comment]
I added some code to read the values from A.EXE in commit 08fb4834785773f697a4230d00cb9c99b622093c. Here are some of them:
box1a: 0, 12, 25, 37, 50, 62, 75, 87
box1b: 0, 16, 32, 48, 100, 200, 300, 400
...
box3b: 64, 72, 80, 88, 96, 104, 112, 120
box4: 0, 8, 16, 24, 32, 40, 48, 56, 0, 16, 32, 48, 100, 200, 300, 400

I experimented with the 5 thickness bits, 3 height bits of raised platforms a few weeks ago but I didn't get anywhere with them.

**Comment by Carmina16** [comment]
It should be `0tttthhh`.

0, 12, ... 87; 0, 16, ... 400; 0, 16, ... 112 are heights, the third array is for the wilderness, the selection of other 2 depends on a flag I haven't investigated yet.

6, 13, ... 100 and 0, 8 ... 120 should be thickness for inside/outside

Those are used in collision calculations, so the actual values used in the rendering may differ.

**Comment by Allofich** [comment]
Carmina16, thank you for the response.

If I may ask you another question, in the linked discussion, you said

> It seems they point on the first element.

Does this mean `*BOXCAP 3` and `*BOXSIDE 0` point to the first element (image resource) of `walla.set`?
Is this a response to "How are the five elements of walla.set assigned to *BOXCAP 3 and *BOXSIDE 0?" Does it mean that in `*BOXCAP 3`, `*BOXCAP` points to the start of `walla.set`, and `3` is the index of `walla.set` to use? (Sorry if afritz1 had no trouble understanding and this is already solved)

Also it seems to me you did not answer his question 

> should be interpreted like this?

Maybe you just did not have the information yet to answer?

Your reverse-engineering results are very valuable, but I'm just a little worried some might not be used  because of miscommunication. It may just be me unable to understand because of lack of knowledge, though.

**Comment by afritz1** [comment]
The .INF parser is a little complex, but the part you're talking about is [here](https://github.com/afritz1/OpenTESArena/blob/83fd18aa05e99285d9114fab2eb1feaff563a54d/OpenTESArena/src/Assets/INFFile.cpp#L545) between lines 545 and 628. `currentIndex` points to the first index in a .SET.

`WallState` definition is [here](https://github.com/afritz1/OpenTESArena/blob/83fd18aa05e99285d9114fab2eb1feaff563a54d/OpenTESArena/src/Assets/INFFile.cpp#L43). The texture index of the first image in a .SET is what each boxcap and boxside points to (like @Carmina16 said: "the first element"). For example, in `CRYSTAL3.INF` (the one in the `ARENA` directory):

```
@FLATS
*BOXCAP 0
*CEILING 135 346
floora.set      #4
*BOXCAP 1
floorb.set      #2
*BOXCAP 2
floorc.set      #3
...
```

`floora.set` has four images with texture indices 0, 1, 2, and 3. `floorb.set` has two images with texture indices 4 and 5, etc.. The texture indices are based on the position in the .INF file so each image gets a unique identifier.

`*BOXCAP` is an array of indices [here](https://github.com/afritz1/OpenTESArena/blob/216d16a994f406e4a572d5a8714d50dedd21469e/OpenTESArena/src/Assets/INFFile.h#L109). `*BOXCAP 0` will be 0, `*BOXCAP 1` will be 4, etc..

Each line starting with an asterisk is put on the stack until they are all popped by a non-asterisk line.

On another note, I have a feeling that `*DOOR` lines actually do need to be read, because otherwise there doesn't appear to be any data telling which kind of door a door voxel is (swinging, raising, sliding, or splitting).

> Your reverse-engineering results are very valuable, but I'm just a little worried some might not be used because of miscommunication.

This is part of the reason why I gave them write access to the wiki, because some of the important details can be lost/forgotten if they're only written on commits.

**Comment by Allofich** [comment]
OK, well I'd need to take time to study the code to truly understand that, but at least it looks like the needed information about the .SET files was understood.

**Comment by Carmina16** [comment]
Door type is stored in two higher bits of the lower byte, and the sound slots corresponding to the door type are hardcoded: 6, 14 and 15.

**Comment by Carmina16** [comment]
How does that seam occur? There is no ceiling over those walls or it is not bright enough?

**Comment by Allofich** [comment]
Hi afritz1, thank you for your work but this is not fixed. You can tell by comparing the textures.

First, the wall in current master.

![1](https://user-images.githubusercontent.com/19624336/40303980-df0c4968-5d2f-11e8-9c0f-c38ca0190dde.png)

The same section (I moved a bit slightly so the angle isn't exactly the same), with a red line at the top of the wall, and yellow boxes around two identifiable patterns in the texture at the top and bottom of the wall.

![2](https://user-images.githubusercontent.com/19624336/40304014-037866ce-5d30-11e8-8056-1967e98716c5.png)

The same section from original Arena (the same photo I pasted in the original post), with the same pixel patterns in yellow boxes. See that the one at the bottom matches but the one at the top has additional wall texture above it.

![3](https://user-images.githubusercontent.com/19624336/40304053-2c499db6-5d30-11e8-911c-8afa0f36ecab.png)

**Comment by Allofich** [comment]
> How does that seam occur? There is no ceiling over those walls or it is not bright enough?

There is no ceiling there, in OpenTESArena. This wall can be seen by loading Fortress of Ice and looking to the east at the start location.

**Comment by Allofich** [comment]
By the way, it seems to be the border wall on the automap (grey squares), which may be relevant.

**Comment by Allofich** [comment]
I suppose since you said "there's still a small gap at the top of the platform" you know already about the gap and maybe those screenshots I posted weren't necessary. But they show that the gap does not exist in the original game, so the issue of "Low wall in Fortress of Ice" isn't fixed, even relevant data is being correctly gotten now.

Maybe border walls are always set to be a full, floor-to-ceiling height wall regardless of the data in the voxel.

**Comment by Allofich** [comment]
Sorry, I compiled the latest source and saw no change in the walls there, but just to be sure I re-ran CMake and rebuilt again, and now I'm seeing different results. I posted several things here about this not being fixed but I retract that, it looks like it probably is.

**Comment by Allofich** [comment]
> How does that seam occur? There is no ceiling over those walls

Yes, in OpenTESArena. Probably also in original Arena. I compared the wall textures and they appear to match, ending on the same pixel of the texture. I think in original Arena it is just more difficult to notice the gap because of the low screen resolution. I will try to take a video, it may be more noticeable with movement.

**Comment by Allofich** [comment]
[WallHeight.zip](https://github.com/afritz1/OpenTESArena/files/2022348/WallHeight.zip)

Here you can see, there does seem to be a gap in the original game. I moved the cursor around the area where I could notice it.

Also some Snow Wolves started shooting spells at me at the end, which lit up the area which helps to see.

**Comment by Carmina16** [comment]
In the original game, the ceiling, floor, and walls are drawn independently, so they will always be looking differently from the real 3D objects.

**Comment by Allofich** [comment]
I guess you mean the part at around 0:13 where it looks a bit a gap and I put the mouse cursor there? At about 0:19, though, there definitely seems to be a visible gap at the top of the wall, visible in the corner.
